File: crack-data-structures-and-algorithms/leetcode/letter_combinations_of_a_phone_number_q17.cpp
#include <string>
#include <string_view>
#include <vector>

using namespace std;

// 核心思路
// 回溯遍历：对于当前digit对应的字母集合逐一生成一个串，然后递归处理下一个digit
// 生成下一个字母的串前需要把上一个字母去掉（对应 L35~L36）
// 递归结束条件为整个digits完整遍历结束

constexpr string_view kTable[] = {
    "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"
};

class Solution {
public:
    vector<string> letterCombinations(string digits) {
        if (digits.empty()) {
            return {};
        }

        vector<string> result;
        string combi;
        gen_combination(digits, 0, combi, result);
        return result;
    }

    void gen_combination(const string& digits, int pos, string& combi, vector<string>& result) {
        if (pos == digits.size()) {
            result.push_back(combi);
            return;
        }

        auto idx = digits[pos] - '0' - 2;
        for (char ch : kTable[idx]) {
            combi += ch;
            gen_combination(digits, pos + 1, combi, result);
            combi.pop_back();
        }

        return;
    }
};

File: crack-data-structures-and-algorithms/leetcode/python-impl/letter_combinations_of_a_phone_number_q17.py
# 核心思想：
# 每个数字都对应几个字母，递归往右拼接生成
# 递归基础条件：i == len(digits)，即已经检查完毕
class Solution(object):
    def letterCombinations(self, digits):
        """
        :type digits: str
        :rtype: List[str]
        """
        table = {
            '2': 'abc',
            '3': 'def',
            '4': 'ghi',
            '5': 'jkl',
            '6': 'mno',
            '7': 'pqrs',
            '8': 'tuv',
            '9': 'wxyz'
        }

        results = []
        combi(digits, 0, table, '', results)
        return results


def combi(digits, i, table, s, results):
    if i == len(digits):
        if s != '':
            results.append(s)
        return

    digit = digits[i]
    for c in table[digit]:
        combi(digits, i + 1, table, s + c, results)
